### DISCUSSION TITLE
"Uncaught Error: Dynamic require of "stream" is not supported" - Electron Vite

### DISCUSSION THREAD

**Discussion by troy-nebula** [serialport/node-serialport#2657] [Q&A]
**SerialPort Version**
11.0.0

**Node Version**
18.12.1

Electron Version
24.4.0

**Platform**
Windows 11

**Architecture**
x64

**Hardware or chipset of serialport**
N/A

**What steps will reproduce the bug?**

1. Create a project with Electron-Vite (React + TS).
2. `npm install serialport`
3. Use this:
```
import { SerialPort } from 'serialport'

const PORT = "COM3";
const serialPort = new SerialPort({
  path: PORT,
    baudRate: 115200,
    autoOpen: false,
}).setEncoding("utf8");
```
4. `npm run dev`
5. Console output shows:
```
caught Error: Dynamic require of "stream" is not supported
    at chunk-AUZ3RYOM.js?v=a13c3e57:12:9
    at node_modules/.vite-electron-renderer/stream.mjs (stream.mjs:1:50)
    at __init (chunk-AUZ3RYOM.js?v=a13c3e57:15:56)
    at node_modules/@serialport/parser-byte-length/dist/index.js (index.js:4:18)
    at __require2 (chunk-AUZ3RYOM.js?v=a13c3e57:18:50)
    at node_modules/serialport/dist/index.js (index.js:17:14)
    at __require2 (chunk-AUZ3RYOM.js?v=a13c3e57:18:50)
    at index.js:28:46
```

**Additional information**
I'm new to this, but seems like Vite doesn't like Serialport's use of `require("stream");`. Is it possible to use Serialport with Vite? I've been trying for a couple of days and haven't made any progress.

**Comment by troy-nebula** [comment]
For anyone looking, I'm sorry to say I don't know how I fixed it. After tinkering with different permutations of frameworks to find a pattern in which had a similar error and which worked, seemingly spontaneously the problem evolved into one with an error of `native build was found for platform=win32 arch=x64 runtime=electron`.

The solution to that was to use `const { SerialPort } = require('serialport')` instead of `import { SerialPort } from 'serialport';`.

If it's obvious to anyone why, or what happened, I'd appreciate your insight.

**Reply by GazHank** [reply]
There is a very brief mention of the change you identified in the upgrade guide https://serialport.io/docs/guide-upgrade#serialport-package 

```
const { SerialPort } = require('serialport')
// or for esm
import { SerialPort } from 'serialport'
```

Perhaps this could be better sign posted?  I'm sure this discussion item with help others find a fix, but if you would like to suggest a change to some of the docs to make this easier for others to find then please feel free to propose a change to the serialport.io website https://github.com/serialport/website

**Reply by rotu** [reply]
@troy-nebula your workaround that also worked for me:
```js
import { createRequire } from 'node:module';
const { SerialPort } = createRequire(import.meta.url)("serialport")
```

It also worked to do this:
```js
const { SerialPort } = await import(import.meta.resolve("serialport"))
```

**Comment by Przeblysk** [comment]
same problem
